### ISSUE TITLE
Capture file system changes during build

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#11837]
Currently when VFS retention is enabled we only capture file system changes between builds. This means that if the user changes one of the source files after it has been processed by the build, but before the build finishes, the next build will ignore the change.

We can do better than this. The idea is to keep watching the file system during the build, and ignore irrelevant events, such as those related to outputs currently being generated. We can also watch only "primary inputs", i.e. inputs that are never generated by the build.

Problems we want to address:

1. Pick up changes that happen to primary inputs after the build started but before the task executed.
2. Pick up changes that happen to input of a task while it is executing and fail the build.
3. Pick up changes that happen to inputs after the task finished
4. Improve performance of registering watchers by doing that during the build
5. Detect if tasks change outputs of other tasks without declaring them
6. Pick up changes that happen to outputs of tasks after they have finished

Item 1) and 3) are the most important thing, since they affect correctness, aka changes are not picked up. 2) also affects correctness though that is a problem we already have. 5) would be nice, though it is also something we don’t necessarily have to do now.

Design: https://docs.google.com/drawings/d/1K8xoU__dumqgoM5u3is21SR5qMbSokNHDS2bx61Eh8s/

---
cc: @gradle/build-cache

**Comment by lptr** [comment]
Duplicate of #11901.

**Comment by wolfs** [comment]
Keeping the watchers alive during the build has only a performance benefit if we don't do anything else. So this issue depends on #11901 but isn't a duplicate.

**Comment by wolfs** [comment]
This is the design doc for the necessary changes: https://docs.google.com/document/d/15sJC1vDwTwOy5W4-jliNbWENkbvrT54mQJV13qLulWg/edit#heading=h.34w1wa8wko9o

**Comment by lptr** [comment]
I've converted this back to an issue, and closing it.
